### DISCUSSION TITLE
Why is font rendering for line text so bad?

### DISCUSSION THREAD

**Discussion by rcdailey** [plantuml-stdlib/C4-PlantUML#347] [Q&A] [answered]
All C4 diagrams have really weird spacing issues with text on lines. See screenshot below. I've tried different fonts but haven't had much luck on kerning/spacing between letters in words. It's so bad that sometimes it looks like a line of spaced out letters instead of words.

![image](https://github.com/plantuml-stdlib/C4-PlantUML/assets/1768054/2141b0eb-4b74-4e55-8d3e-d13e31d0ea0e)

I'm using the public Plant UML server for rendering diagrams. I was able to work around the issue by putting custom fonts on a dockerized version of PlantUML, but that won't work for my github workflow since I do not host my own server, nor do I want to.

What is the fix for the font issue?

**Comment by kirchsth** [comment]
Hi @rcdailey,

I did some trials.
 
I assume you have problems with svg files (and not with png files).
And if you add the following line in your source then the font problem is gone (and you lose layout based on the new width).
```plantuml
skinparam maxMessageSize 0
```
like e.g. 

```plantuml
@startuml
!include <C4/C4_Sequence>

skinparam maxMessageSize 0

Rel(a,a, "Perform the usual session auth")
@enduml
```
[![](https://www.plantuml.com/plantuml/svg/3Ssn2i8m40RW_PxYxA92GOUE8eNEWjW7aAFzjS7aghaSY4zlzW--tehauYdI9kWKVGQVXkumTFSH7uTEE1DjsLv1tv8bSPBl1MRon1X-u2FH3N4dhRJSN94VQqvS5h2RIsIhDApAucLfzjH3vthz0G00)](https://www.plantuml.com/plantuml/uml/3Ssn2i8m40RW_PxYxA92GOUE8eNEWjW7aAFzjS7aghaSY4zlzW--tehauYdI9kWKVGQVXkumTFSH7uTEE1DjsLv1tv8bSPBl1MRon1X-u2FH3N4dhRJSN94VQqvS5h2RIsIhDApAucLfzjH3vthz0G00)

[![](https://www.plantuml.com/plantuml/svg/3Sr12i8m40NG_PnYx4gXe8ikHGfT2s8F89_sQuD9gfaCYAVNTu7NMqMfdg9iGfwYpzJZqEs7xZRoxSmJJoBs3Fc5WgG9dpFDyE0OljI3o9LnYnQjDXUM-rgIreNgveXgD0jhLdXTcftqpFD_-W40)](https://www.plantuml.com/plantuml/uml/3Sr12i8m40NG_PnYx4gXe8ikHGfT2s8F89_sQuD9gfaCYAVNTu7NMqMfdg9iGfwYpzJZqEs7xZRoxSmJJoBs3Fc5WgG9dpFDyE0OljI3o9LnYnQjDXUM-rgIreNgveXgD0jhLdXTcftqpFD_-W40)

Are my assumptions correct?

I think the problem is in the svg rendering of plantuml: I `maxMessageSize` is set (0 deactivates it) all words are positioned separately and therefore it doesn't match so good anymore (I assume they make a rounding of 0.5 but this has to be checked on PlantUML side).

I overlaid both diagrams in one svg file and then it's simpler to see the differences between the files
(second line should be ok for you).

![SpaceProblems_overlay](https://github.com/plantuml-stdlib/C4-PlantUML/assets/21959385/d1218816-7cf0-47fe-871d-a18ffe67b528)

I entered an issue in the [forum](https://forum.plantuml.net/18858/exports-diagrams-maxmessagesize-wrapwidth-spacing-issues), maybe they can change the PlantUML implementation (remove rounding or write all words combined).

Best regards
Helmut

**Reply by rcdailey** [reply]
I am doing PNG in my case

**Comment by kirchsth** [comment]
Do you get better results with `skinparam maxMessageSize 0`?

**Reply by rcdailey** [reply]
I tested the following example C4 sequence diagram:

```plantuml
@startuml
!include https://raw.githubusercontent.com/plantuml-stdlib/C4-PlantUML/master/C4_Sequence.puml

skinparam maxMessageSize 0

Container(c1, "Single-Page Application", "JavaScript and Angular", "Provides all of the Internet banking functionality to customers via their web browser.")

Container_Boundary(b, "API Application")
  Component(c2, "Sign In Controller", "Spring MVC Rest Controller", "Allows users to sign in to the Internet Banking System.")
  Component(c3, "Security Component", "Spring Bean", "Provides functionality Related to signing in, changing passwords, etc.")
Boundary_End()

ContainerDb(c4, "Database", "Relational Database Schema", "Stores user registration information, hashed authentication credentials, access logs, etc.")

Rel(c1, c2, "Submits credentials to", "JSON/HTTPS")
Rel(c2, c3, "Calls isAuthenticated() on")
Rel(c3, c4, "select * from users where username = ?", "JDBC")

SHOW_LEGEND()
@enduml
```

It appears like this:

![image](https://github.com/plantuml-stdlib/C4-PlantUML/assets/1768054/153ff400-3abd-4ed2-ad75-2a7c4daa7fe4)

If I disable `skinparam`, it looks like this:

![image](https://github.com/plantuml-stdlib/C4-PlantUML/assets/1768054/bf63f901-5fc5-47da-b5f6-31c1be31afa2)

There's not much of a difference, although it does appear that the first one looks a little better.

Neither of the above renders looks a good as the version I get when I use a local plantuml instance:

![image](https://github.com/plantuml-stdlib/C4-PlantUML/assets/1768054/c6e3d59c-3b67-4758-bf47-076474d9f436)

**Comment by kirchsth** [comment]
If I understand you correctly then 
- local version offers the expected output and 
- only the online server has the problems (and online with `skinparam maxMessageSize 0` is better).

a) Is your suggested solution is a different font?
b) Did you found a better font?

**Reply by kirchsth** [reply]
A possible solution could be to use a rendering hint in PlantUML. 

```java
g2d.setRenderingHint(RenderingHints.KEY_FRACTIONALMETRICS, RenderingHints.VALUE_FRACTIONALMETRICS_ON);
```

If this is used then it creates it a text like the line 3.

![SampleWithExactXValues](https://github.com/plantuml-stdlib/C4-PlantUML/assets/21959385/b596c7af-cbb4-4d21-bf36-d0afafdb348c) 

And the sequence diagram would look like 
![Test______Rendering_C4_Sequence](https://github.com/plantuml-stdlib/C4-PlantUML/assets/21959385/5a0c2f5a-1fbe-495a-a289-7380bfef1423)

**Reply by kirchsth** [reply]
Following [Plantuml PR](https://github.com/plantuml/plantuml/pull/1843) is created

**Reply by kirchsth** [reply]
Hi @rcdailey,
can you please check the new generated sequence diagram?
Is it better?
Thank you and best regards
Helmut

**Reply by kirchsth** [reply]
In previous diagram I compared the online and local version. Following image shows the real difference between the different local jar versions:
- the 3. line is calculated via local jar with **new** implementation
- the 4. line is calculated via local jar with **old** implementation
  
  
![SampleWithExactXValues](https://github.com/plantuml-stdlib/C4-PlantUML/assets/21959385/d446871c-6161-42e1-a223-21c5e0f3eebd)

**Comment by kirchsth** [comment]
Hi @rcdailey,

[PlantUML snapshot ~v1.2024.6beta12](https://github.com/plantuml/plantuml/releases/tag/snapshot) includes MR [set RenderingHints.VALUE_FRACTIONALMETRICS_ON](https://github.com/plantuml/plantuml/pull/1843).

Can you test/compare the snapshot output with your old results and write some feedback if it looks better or not

Thank you and best regards
Helmut

**Reply by rcdailey** [reply]
HI there, I did get your last message but have been swamped this week. I will absolutely test this for you as soon as I can. I have some free time today so I'll try to get it done. Thank you very much for personally handling this for me!

EDIT: I actually remembered that all of this was done on my work PC, so I probably won't be able to try it until Monday at the earliest.

If in the meantime you can share some steps to test that might help -- if the plantuml public server is already updated and all I have to do is use that, then no problem!

**Reply by kirchsth** [reply]
Hi @rcdailey

atm the online server uses only v1.2024.5.

But if you have Java installed on your machine, and you download e.g. the [https://github.com/plantuml/plantuml/releases/tag/snapshot plantuml-mit-SNAPSHOT.jar](https://github.com/plantuml/plantuml/releases/download/snapshot/plantuml-mit-SNAPSHOT.jar) and  [https://github.com/plantuml/plantuml/releases/tag/v1.2024.5 plantuml-mit-1.2024.5.jar](https://github.com/plantuml/plantuml/releases/download/v1.2024.5/plantuml-mit-1.2024.5.jar) files from the different releases
then you can simply generate the old/new version with following commands in 2 different folders and compare e.g. your typically used *.png output files.  (I used in my sample a C4-PlantUML/percy file, please update the *.puml path with your files)

```bash
java -jar plantuml-mit-1.2024.5.jar -o ./oldVersion "./C4-PlantUML/percy/C4_Sequence Diagram Sample - bigbankplc.puml"

java -jar plantuml-mit-SNAPSHOT.jar -o ./newVersion "./C4-PlantUML/percy/C4_Sequence Diagram Sample - bigbankplc.puml"
```

more command details see https://plantuml.com/command-line.

E.g. on my machine I got following output:

![image](https://github.com/plantuml-stdlib/C4-PlantUML/assets/21959385/6fdf21d5-d296-4398-b843-044a04926d03)

BR Helmut

**Reply by kirchsth** [reply, marked as answer]
In the meantime the online server is updated to v1.2024.6, so you can make all your (png) checks with the online server too.
BR Helmut

[![](https://www.plantuml.com/plantuml/png/DSon2i8m48VnFKznTb8ek7GKANHpaYeuofdyfO4anTpT-ziMkd-_hnVbgfQJEyJYamNGPUZEG_T-u6Se7bVdHgGZjznISqVzpZMJJY0JuqG2aJWNOjEfEUtjPsd7TI1AjzUJ5hGLA1ezAyBkLk9wb60v_G40)](https://www.plantuml.com/plantuml/uml/DSon2i8m48VnFKznTb8ek7GKANHpaYeuofdyfO4anTpT-ziMkd-_hnVbgfQJEyJYamNGPUZEG_T-u6Se7bVdHgGZjznISqVzpZMJJY0JuqG2aJWNOjEfEUtjPsd7TI1AjzUJ5hGLA1ezAyBkLk9wb60v_G40)
